File: DSA/Vikash_kumar/Day_6_Question_1.py
#Day6-Question1-
class ListNode:
    def __init__(self, value=0, next=None):
        self.value = value
        self.next = next

def has_cycle(head):
    if not head:
        return "no cycle"
    
    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        
        if slow == fast:
            return "cycle"
    
    return "no cycle"

def create_linked_list(values, pos):
    if not values:
        return None
    
    head = ListNode(values[0])
    current = head
    nodes = [head]

    for value in values[1:]:
        new_node = ListNode(value)
        current.next = new_node
        current = new_node
        nodes.append(new_node)

    if pos != -1 and pos < len(values):
        current.next = nodes[pos]  # Create the cycle

    return head

values = list(map(int, input("Enter the node values (space-separated): ").split()))
pos = int(input("Enter the position to create a cycle (-1 for no cycle): "))

# Create the linked list
head = create_linked_list(values, pos)

# Check for cycle
result = has_cycle(head)
print(result)

File: DSA/RAJAS-BAGGA/Day_6_Question_1.java
class Node {
    int data;
    Node next;
    Node(int data) {this.data = data;}
}

public class Day_6_Question_1 {
    public static String hasCycle(Node head) {
        Node slow = head;
        Node fast = head;
        
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow)
                return "cycle";
        }
        return "no cycle";
    }
    public static void main(String[] args) {
        Node n1 = new Node(1);
        n1.next = new Node(2);
        n1.next.next = new Node(3);
        n1.next.next.next = new Node(4);
        n1.next.next.next.next = n1.next;
        
        System.out.println(hasCycle(n1));
    }
}
